octopus_sensing/tests/test_camera_streaming.py

```diff
@@ -0,0 +1,130 @@
+import os
+import time
+import tempfile
+import queue
+import threading
+import pytest
+
+import octopus_sensing.devices.camera_streaming as camera_streaming
+from octopus_sensing.common.message_creators import start_message, stop_message, terminate_message
+
+class MockedCv2Module:
+
+    def VideoCapture(self):
+        return MockedVideoCaptureModule()
+    
+    def VideoWriter_fourcc(self, a, b, c, d):
+        return ('XVID')
+    
+    def VideoWriter(self, file_name, codec, fps, _video_size):
+        return MockedVideoWriterModule(file_name)
+
+    '''
+    Get the information of your computer using:
+    >>> import cv2
+    >>> cap=cv2.VideoCapture(0)
+    >>> print(cap.get(cv2.CAP_PROP_FRAME_WIDTH))
+    >>> print(cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
+    '''
+    def CAP_PROP_FRAME_WIDTH(self):
+        return (float(640.0))
+    
+    def CAP_PROP_FRAME_HEIGHT(self):
+        return (float(480.0))
+    
+    def imwrite(self, filename, img):
+        a = time.time()
+        # self.output_path/self.name-experiment_id-str(message.stimulus_id).zfill(2)/str(a).jpg
+        open(filename[:-4] + "/" + str(a) + ".jpg", 'w+').write(img)
+
+class MockedVideoCaptureModule:
+
+    def isOpened(self):
+        return True
+
+    def set(self):
+        pass
+
+    def read(self):
+        # bool, 3d matrix
+        return True, [[[0,10,10],[1,11,11]], [[0,10,10],[1,11,11]]]
+    
+    def release(self):
+        pass
+
+
+class MockedVideoWriterModule:
+
+    def __init__(self, file_name):
+        self.file_name = file_name
+
+    def write(self, frame):
+        a = time.time()
+        open(self.file_name[:-4] + "/" + str(a) + ".jpg", 'w+').write(frame)
```

Comment:
The path you're creating here is referring to a directory that doesn't exist. That's why it's failing to write the file. (It gives me a `no such file or directory` error).

It's a better idea to just write to `self.file_name` without modification. So, the test would actually check the file name too. Filenames are important and we want the test to fail if we accidentally changed the file name.

Reply:
Thanks for the patient review!! When I tried the:
`open(self.file_name, 'w+').write(frame)`
It still shows `no such file or directory`, so I was wondering if I'm writing this function in a wrong way. In Camera_streaming.py, the implementation is:
writer.write(frame)
In the unit test, I simply save the image file into the file path, I'm wondering whether it's correct to do so?

Reply:
No worries Yumeng :slightly_smiling_face: 

It seems okay. It should work. I wonder what's happening. You can print the `self.file_name` and see if it tries to write in a directory that doesn't exist?

Also, if you push your changes, I can run it on my own machine and see what's happening.

Reply:
Sure! I've pushed the changes on [this repo](https://github.com/raynehe/octopus-sensing/blob/unit-test-for-devices/octopus_sensing/tests/test_camera_streaming.py). I'm glad if you could take a look!

When I print the `self.file_name` in `write()` function and run the code, seems that it doesn't show any output in the pytest terminal😦